feat: add flake8-bugbear

jnoortheen opened the pull request:
<!--- Thanks for opening a PR on xonsh! Please include a news entry with your PR
to help keep our changelog up to date! There are instructions available here:
https://xon.sh/devguide.html#changelog -->

<!--- If there is specific issue / feature request that this PR is addressing,
please link to the corresponding issue by using the `#issuenumber` syntax.
Thanks again! -->

## For community
⬇️  **Please click the 👍 reaction instead of leaving a `+1` or 👍  comment**

jnoortheen commented:
>  the question is whether we want to introduce yet another linter.


This contains some important bug detection IMO

jnoortheen commented on amalgamate.py:
```diff
@@ -362,7 +362,7 @@ def format_lazy_import(names):
     """Format a from import line"""
     parts = []
-    for _, module, name, asname in names:
+    for _, module, name, asname in names:  # noqa
```
that module is latter referenced

gforsyth commented:
I'm not opposed. I think bugbear trips up fewer people because the errors are less common than, say, styling issues.

Two asks:
Can this also be added to the precommit config?
Can you remove the pylint bare except directives since they are no longer bare?

jnoortheen commented:
> Two asks:
> Can this also be added to the precommit config?

1. do you want to add flake8 as whole? I think we can add it part of push checks. 

> Can you remove the pylint bare except directives since they are no longer bare?
2. will update it.

jnoortheen commented on xonsh/completers/python.py:
```diff
@@ -211,11 +211,11 @@ def _safe_eval(expr, ctx):
             val = xonsh_safe_eval(expr, builtins.__dict__, transform=False)
             _ctx = builtins.__dict__
-        except:  # pylint:disable=bare-except
+        except Exception:  # pylint:disable=bare-except
```
removed in other places as well.

gforsyth commented:
Yeah, let's add flake8.

Then later we can run precommit itself during the style CI runs and so anyone running precommit locally will be guaranteed to pass on CI

jnoortheen commented on xonsh/inspectors.py:
```diff
@@ -393,7 +393,7 @@ def pdoc(self, obj, oname=""):
                 lines.append(head("Init docstring:"))
                 lines.append(indent(init_ds))
-        elif hasattr(obj, "__call__"):
+        elif hasattr(obj, "__call__"):  # noqa : use callable
```
here I don't think __call__ and being callable is the same. this wouldn't pass a function and callable would.

daniel-shimon replied:
Functions also have `__call__` set.
Callable's implementation checks for `__call__` itself (or checks the c equivalent).
(see `PyCallable_Check` in http://svn.python.org/projects/python/trunk/Objects/object.c)

jnoortheen replied:
updated

jnoortheen commented:
> Yeah, let's add flake8.
> 
> Then later we can run precommit itself during the style CI runs and so anyone running precommit locally will be guaranteed to pass on CI

added qa as whole to pre-commit hook. Added this as pre-push hook since it will take some time to run.
command to install this hook

```
pre-commit install --hook-type pre-push
```

jnoortheen commented:
what happened to the news-item checking action? it doesn't seem to be running for recent PRs.

gforsyth reviewed (approved):
Changes here look good to me.  @daniel-shimon -- anything else you want to see done here?

As for the news checks, @AstraLuma set them up ages ago and some webhook may not be firing anymore?  
We can probably add a news check with github actions fairly easily.

daniel-shimon commented:
I think it's good - thanks @jnoortheen 👌
